articles/141_static_remapping.md

```diff
@@ -0,0 +1,435 @@
+Because a process can contain multiple nodes, there must be a way to uniquely identify a node in a process.
+This is a feature of ROS 1 remapping.
+
+*Example:*
+
+- Node is an executable at `/bin/my_node`
+- User wants to change `/cat` to `/dog`
+- User types `/bin/my_node /cat:=/dog`
+
+
+### Change the Default Namespace
+The default namespace is the one in which relative names get expanded to.
+This should be changeable without affecting FQN.
+ROS 1 has this feature using either the environment variable `ROS_NAMESPACE` or the argument `__ns`.
+
+*Example:*
+
+- Node is in namespace `/ns` and uses relative name `bar`
+- User changes the default namespace to `/foo`
+- The final name is `/foo/bar`
+
+### Change the Node Name
+The node name is used in log messages and to create private names.
+ROS 1 has this feature using the argument `__node`.
+
+*Example:*
+
+- Node is named `camera_driver` and uses a private name `camera_info`
+- User changes the node name to `left_camera_driver`
+- The final name is `/ns/left_camera_driver/camera_info` and log messages use `left_camera_driver`
+
+## Remapping Names in ROS 1
+Remapping is a feature that also exists in ROS 1.
+In ROS 1 remapping works by passing in [arguments](http://wiki.ros.org/Remapping%20Arguments) to each node.
+Client libraries also have APIs in code to pass remapping rules when the node is initialized.
+A remap rule consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+Both sides of a rule are [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b).
+Before a name is remapped it is also [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f).
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+
+## Remapping rule syntax
+This is a proposal for the ROS 2 remapping rule syntax.
+It attempts to be the same as ROS 1 syntax when possible.
+
+Use cases supported by this syntax:
+
+- Remap One Node in a Process
+- Change a Namespace
+- Change a Basename
+- Exact FQN Replacement
+- Exact Relative Name Replacement
+- Remap via Command Line
+- Change the Default Namespace
+- Change the Node Name
+
+Not supported:
+
+- Change a Token
+- Pre-FQN Remapping
+
+### How the Syntax Works
+
+The structure of a remapping rule is `match:=replacement`.
+`match` tests if a name should be remapped.
+`replacement` says what the new name will be.
+`:=` behaves the same as it does in ROS 1.
+
+Example rules are:
+
+- `foo:=bar`
+- `/foo/bar:=fiz/buzz`
+- `nodename:~/foo:=foo`
+- `**/foo:=\1/bar`
+
+#### Match Part of a Rule
+
+The match part of a rule uses these operators:
+
+- `*` matches a single token
+- `**` matches zero or more tokens delimited by slashes
+- `~` is replaced with the node's name anywhere it is used
```

Comment:
> What do you think about removing ~ entirely in favor of {node}?

(we already talked about this in person since @sloretz asked here, but to summarize...)

Yeah, I believe that was discussed at some point, but the consensus (IIRC) was to leave it since it follows the filesystem metaphor and because it exists in ROS 1. The compromise was to explicitly disallow the `~foo` style syntax that ROS 1 allowed, but that differed from how most systems treat it in a filesystem path. There might be some details here: https://github.com/ros2/design/pull/90#r78666048 and perhaps other locations in that pr.